### PR TITLE
How to build a MSI file locally.

### PR DESCRIPTION
**Pull request by gounthar**
- [X] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [X] Ensure that the pull request title represents the desired changelog entry
- [X] Please describe what you did
- ~~Link to relevant issues in GitHub or Jira~~
- ~~Link to relevant pull requests, esp. upstream and downstream changes~~
- ~~Ensure you have provided tests - that demonstrates feature works or fixes the issue~~

I have written a quick howto about how to build locally a MSI file.

### CODE REVIEW

File: msi/docs/build-msi-locally.md [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+You can install [MSBuild](https://aka.ms/vs/17/release/vs_BuildTools.exe) from Visual Studio or from the [Build Tools for Visual Studio](https://visualstudio.microsoft.com/downloads/#build-tools-for-visual-studio-2022).
+
+This command line tool is used to build the MSI file.
+
+## Install .NET Framework 3.5
+
+You may already have it installed on your machine, but not activated.
+You can activate it from the Windows Features dialog box.
+
+To access this dialog box, smash the `WinKey` plus `R`, then enter the command `appwiz.cpl` and push enter.
```
Comment (NotMyFault) [inline msi/docs/build-msi-locally.md:37]:
```suggestion
To access this dialog box, smash `<kbd>⊞ Win</kbd> + <kbd>R</kbd>, then enter the command `appwiz.cpl` and push enter.
```
GitHub highlights the keyboard element with a specific button-like design.

File: msi/docs/build-msi-locally.md [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+
+Mode                 LastWriteTime         Length Name
+----                 -------------         ------ ----
+-a----        01/12/2022     20:53      105107456 jenkins-2.380.msi
+-a----        01/12/2022     20:53             84 jenkins-2.380.msi.sha256
+```
+
+
+
```
Comment (NotMyFault) [inline msi/docs/build-msi-locally.md:139]:
```suggestion
```

**Review by NotMyFault — COMMENTED** [review]
Thanks for this detailed step-by-step guide!

Left some optional feedback for consideration.

File: msi/docs/build-msi-locally.md (hunk 1 above)
Comment (lemeurherve) [reply msi/docs/build-msi-locally.md:37]: Not sure if there is a better display on your side but here is how this square character is rendered on Mac (seems unrecognized?):
image

Also, in your suggestion the first backtick should be removed ;)

File: msi/docs/build-msi-locally.md [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+# Pre-requisites
+
+## Jenkins WAR file
+
+First of all, you should get the Jenkins war file that will be inside that MSI file.
+You can get it from the official Jenkins website or from the Jenkins update center.
+I will use the official Jenkins website in this article.
+
+Check [Jenkins download page](https://www.jenkins.io/download/) and download the latest weekly version of Jenkins for example.
+For me, it is [jenkins.war](https://get.jenkins.io/war/2.380/jenkins.war) version 2.380.
```
Comment (lemeurherve) [inline msi/docs/build-msi-locally.md:12]:
```suggestion
Currently it is [jenkins.war](https://get.jenkins.io/war/2.380/jenkins.war) version 2.380.
```

File: msi/docs/build-msi-locally.md [hunk 4]
```diff
@@ -0,0 +1,139 @@
+Should you ever need to rebuild a Jenkins MSI on your Windows machine, here is a way to do it.
+
+# Pre-requisites
+
+## Jenkins WAR file
+
+First of all, you should get the Jenkins war file that will be inside that MSI file.
+You can get it from the official Jenkins website or from the Jenkins update center.
+I will use the official Jenkins website in this article.
```
Comment (lemeurherve) [inline msi/docs/build-msi-locally.md:9]: It's a doc not an article or a blog post, not sure I'd use "I" and this formulation. Idem for the following "for me".
(Subjective)

File: msi/docs/build-msi-locally.md [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+## Git tool
+
+You have a few options here:
+* Install [Git for Windows](https://community.chocolatey.org/packages/git)
+* Use your IDE to clone the repository
+* Use [GitHub Desktop](https://desktop.github.com/)
+* Use [SourceTree](https://www.sourcetreeapp.com/)
+* Use Cygwin integrated [Git command](https://cygwin.com/packages/summary/git.html)
+* [...]
```
Comment (lemeurherve) [inline msi/docs/build-msi-locally.md:24]: Off topic? I would simply specify git as requirement and a link to the official website.

**Review by lemeurherve — COMMENTED** [review]
Some remarks, looks good otherwise!

File: msi/docs/build-msi-locally.md (hunk 1 above)
Comment (NotMyFault) [reply msi/docs/build-msi-locally.md:37]: Looks correct. It's unicode, similar to [⌘](https://www.compart.com/en/unicode/U+2318).

File: msi/docs/build-msi-locally.md [hunk 6]
```diff
@@ -34,7 +34,7 @@ This command line tool is used to build the MSI file.
 You may already have it installed on your machine, but not activated.
 You can activate it from the Windows Features dialog box.
 
-To access this dialog box, smash the `WinKey` plus `R`, then enter the command `appwiz.cpl` and push enter.
+To access this dialog box, smash `<kbd>⊞ Win</kbd> + <kbd>R</kbd>, then enter the command `appwiz.cpl` and push enter.
```
Comment (lemeurherve) [inline msi/docs/build-msi-locally.md:37]:
```suggestion
To access this dialog box, smash <kbd>⊞ Win</kbd> + <kbd>R</kbd>, then enter the command `appwiz.cpl` and push enter.
```

File: msi/docs/build-msi-locally.md (hunk 4 above)
Comment (gounthar) [reply msi/docs/build-msi-locally.md:9]: You're right.

File: msi/docs/build-msi-locally.md (hunk 5 above)
Comment (gounthar) [reply msi/docs/build-msi-locally.md]: Done. Thanks.

File: msi/docs/build-msi-locally.md [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+You can install [MSBuild](https://aka.ms/vs/17/release/vs_BuildTools.exe) from Visual Studio or from the [Build Tools for Visual Studio](https://visualstudio.microsoft.com/downloads/#build-tools-for-visual-studio-2022).
+
+This command line tool is used to build the MSI file.
+
+## Install .NET Framework 3.5
+
+You may already have it installed on your machine, but not activated.
+You can activate it from the Windows Features dialog box.
+
+To access this dialog box, smash <kbd>⊞ Win</kbd> + <kbd>R</kbd>, then enter the command `appwiz.cpl` and push enter.
```
Comment (lemeurherve) [inline msi/docs/build-msi-locally.md:30]:
```suggestion
To access this dialog box, press the keys <kbd>⊞ Win</kbd> + <kbd>R</kbd>, then enter the command `appwiz.cpl` and push enter.
```

File: msi/docs/build-msi-locally.md [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+
+There are quite a few ways to get Git on Windows, but the most straightforward is to see what the [official Git website recommends](https://git-scm.com/download/win).
+
+## Install MSBuild
+
+You can install [MSBuild](https://aka.ms/vs/17/release/vs_BuildTools.exe) from Visual Studio or from the [Build Tools for Visual Studio](https://visualstudio.microsoft.com/downloads/#build-tools-for-visual-studio-2022).
+
+This command line tool is used to build the MSI file.
+
+## Install .NET Framework 3.5
```
Comment (lemeurherve) [inline msi/docs/build-msi-locally.md:25]:
```suggestion
### Install .NET Framework 3.5
```

File: msi/docs/build-msi-locally.md [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+
+> Turn Windows features on or off.
+
+Tick the `.NET Framework 3.5` entry and install.
+
+Important: now run Windows Update to check for security updates.
+
+If it is not installed yet, you can install [.NET Framework 3.5](https://dotnet.microsoft.com/en-us/download/dotnet-framework/net35-sp1) from the Windows Features.
+
+## Check you have Powershell 5.1
```
Comment (lemeurherve) [inline msi/docs/build-msi-locally.md:41]:
```suggestion
### Check if you have Windows PowerShell 5.1
```

Comment (lemeurherve) [reply msi/docs/build-msi-locally.md:41]: What if they have PowerShell 7.x? Does it change anything?

FTR: https://learn.microsoft.com/en-us/powershell/scripting/whats-new/differences-from-windows-powershell?view=powershell-7.3

File: msi/docs/build-msi-locally.md [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+
+# Clone the Jenkins repository and build the MSI
+
+## Clone the Jenkins packaging repository
+
+Choose your `git` tool and clone the [Jenkins packaging repository](https://github.com/jenkinsci/packaging.git) on your machine.
+
+## Prepare the build
+
+Open a `terminal` and go to the folder where you cloned the repository. For me it is `C:\dev\jenkins\git\ci\packaging\`.
```
Comment (lemeurherve) [inline msi/docs/build-msi-locally.md:53]:
```suggestion
Open a `terminal` and go to the folder where you cloned the repository. For example `C:\jenkinsci\packaging\`.
```
Simpler path

File: msi/docs/build-msi-locally.md [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+You now have to declare where you downloaded the Jenkins war file so that the build can find it.
```
Comment (lemeurherve) [inline msi/docs/build-msi-locally.md:54]:
```suggestion
You now have to declare where you downloaded the Jenkins war file so the build can find it.
```

File: msi/docs/build-msi-locally.md [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+
+```powershell
+$env:War = "$env:USERPROFILE\jenkins.war"
+```
+
+If you ever moved it into your repository clone folder, you can use this command instead:
+
+```powershell
+$env:War = "C:\dev\jenkins\git\ci\packaging\msi\build\jenkins.war"
```
Comment (lemeurherve) [inline msi/docs/build-msi-locally.md:63]:
```suggestion
$env:War = "C:\dev\jenkinsci\packaging\msi\build\jenkins.war"
```

File: msi/docs/build-msi-locally.md [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+
+## Build the MSI
+
+Enter the subfolder `msi\build` and run the following command:
+
+```powershell
+.\build.ps1
+```
+
+For me, this was the output:
```
Comment (lemeurherve) [inline msi/docs/build-msi-locally.md:74]:
```suggestion
You should get an output similar to:
```

File: msi/docs/build-msi-locally.md [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+  ild\packages\WiX.3.11.1\build\..\tools\\WixNetFxExtension.dll -ext C:\dev\jenkins\git\ci\packaging\msi\build\packages\WiX.3.11.1\build\..\tools\\WixUtilExte
+  nsion.dll -ext .\msiext-1.5\WixExtensions\WixCommonUIExtension.dll -ext C:\dev\jenkins\git\ci\packaging\msi\build\packages\WiX.3.11.1\build\..\tools\\WixFir
+  ewallExtension.dll -fv -loc jenkins_en-US.wxl -spdb -contentsfile obj\Release\jenkins.wixproj.BindContentsFileListen-US.txt -outputsfile obj\Release\jenkins.wixproj.BindOutputs
+  FileListen-US.txt -builtoutputsfile obj\Release\jenkins.wixproj.BindBuiltOutputsFileListen-US.txt -wixprojectfile C:\dev\jenkins\git\ci\packaging\msi\build\
+  jenkins.wixproj obj\Release\jenkins.wixobj
+  Windows Installer XML Toolset Linker version 3.11.1.2318
+  Copyright (c) .NET Foundation and contributors. All rights reserved.
+
+  jenkins -> C:\dev\jenkins\git\ci\packaging\msi\build\bin\Release\en-US\jenkins-2.380.msi
+Done Building Project "C:\dev\jenkins\git\ci\packaging\msi\build\jenkins.wixproj" (default targets).
```
Comment (lemeurherve) [inline msi/docs/build-msi-locally.md:104]:
```suggestion
  C:\jenkinsci\packaging\msi\build\packages\WiX.3.11.1\build\..\tools\Light.exe -out C:\jenkinsci\packaging\msi\build\bi
  n\Release\en-US\jenkins-2.380.msi -pdbout C:\jenkinsci\packaging\msi\build\bin\Release\en-US\jenkins-2.380.wixpdb -sw1076 -cultures:en-US -ext C:\S
  upport\users\jenkinsci\packaging\packaging\msi\build\packages\WiX.3.11.1\build\..\tools\\WixUIExtension.dll -ext C:\jenkinsci\packaging\msi\bu
  ild\packages\WiX.3.11.1\build\..\tools\\WixNetFxExtension.dll -ext C:\jenkinsci\packaging\msi\build\packages\WiX.3.11.1\build\..\tools\\WixUtilExte
  nsion.dll -ext .\msiext-1.5\WixExtensions\WixCommonUIExtension.dll -ext C:\jenkinsci\packaging\msi\build\packages\WiX.3.11.1\build\..\tools\\WixFir
  ewallExtension.dll -fv -loc jenkins_en-US.wxl -spdb -contentsfile obj\Release\jenkins.wixproj.BindContentsFileListen-US.txt -outputsfile obj\Release\jenkins.wixproj.BindOutputs
  FileListen-US.txt -builtoutputsfile obj\Release\jenkins.wixproj.BindBuiltOutputsFileListen-US.txt -wixprojectfile C:\jenkinsci\packaging\msi\build\
  jenkins.wixproj obj\Release\jenkins.wixobj
  Windows Installer XML Toolset Linker version 3.11.1.2318
  Copyright (c) .NET Foundation and contributors. All rights reserved.

  jenkins -> C:\jenkinsci\packaging\msi\build\bin\Release\en-US\jenkins-2.380.msi
Done Building Project "C:\jenkinsci\packaging\msi\build\jenkins.wixproj" (default targets).
```

**Review by lemeurherve — COMMENTED** [review]
Additional suggestions 🙂

File: msi/docs/build-msi-locally.md [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+$env:War = "$env:USERPROFILE\jenkins.war"
+```
+
+If you ever moved it into your repository clone folder, you can use this command instead:
+
+```powershell
+$env:War = "C:\dev\jenkins\git\ci\packaging\msi\build\jenkins.war"
+```
+
+## Build the MSI
```
Comment (lemeurherve) [inline msi/docs/build-msi-locally.md:66]:
```suggestion
### Build the MSI
```

File: msi/docs/build-msi-locally.md [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+
+
+Build succeeded.
+    0 Warning(s)
+    0 Error(s)
+
+Time Elapsed 00:00:08.26
+```
+
+## Find the MSI file
```
Comment (lemeurherve) [inline msi/docs/build-msi-locally.md:114]:
```suggestion
## Locate the generated MSI file
```

File: msi/docs/build-msi-locally.md (hunk 9 above)
Comment (gounthar) [reply msi/docs/build-msi-locally.md:41]: I unfortunately have no idea.

File: msi/docs/build-msi-locally.md [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+You now have to declare where you downloaded the Jenkins war file so the build can find it.
+
+```powershell
+$env:War = "$env:USERPROFILE\jenkins.war"
+```
+
+If you ever moved it into your repository clone folder, you can use this command instead:
+
+```powershell
+$env:War = "C:\dev\jenkinsci\packaging\msi\build\jenkins.war"
```
Comment (lemeurherve) [inline msi/docs/build-msi-locally.md:65]:
```suggestion
$env:War = "C:\jenkinsci\packaging\msi\build\jenkins.war"
```
I've missed that `dev`

**Review by lemeurherve — APPROVED** [review]
👏

**Review by NotMyFault — APPROVED** [review]
Thanks!

**Pull request merged**